atf_core/src/atf_core/recorder.py

```diff
@@ -92,21 +103,33 @@ def __init__(self, test):
                 client = SimpleActionClient(action, roslib.message.get_message_class(topic_type))
 
                 # wait for action server
-                client.wait_for_server()
+                if wait_timeout == None:
+                    client.wait_for_server()
+                else:
+                    client.wait_for_server(rospy.Duration(wait_timeout))
                 rospy.loginfo("... action '%s' available.", action)
 
         if test.robot_config != None and 'wait_for_tfs' in test.robot_config:
             listener = tf.TransformListener()
             for root_frame, measured_frame in test.robot_config["wait_for_tfs"]:
+                self.check_for_timeout(start_time, wait_timeout)
                 rospy.loginfo("Waiting for transformation from '%s' to '%s' ...", root_frame, measured_frame)
-                listener.waitForTransform(root_frame, measured_frame, rospy.Time(), rospy.Duration(1))
+                while not rospy.is_shutdown():
+                    self.check_for_timeout(start_time, wait_timeout)
+                    try:
+                        listener.waitForTransform(root_frame, measured_frame, rospy.Time(), rospy.Duration(1))
+                        break # transform is available
+                    except tf2_ros.TransformException as e: #pylint: disable=no-member
+                        pass
+                    rospy.logdebug("... waiting since %.1f sec for transformation from '%s' to '%s' to become available ..."%((rospy.Time.now() - start_time).to_sec(), root_frame, measured_frame))
                 rospy.loginfo("... transformation from '%s' to '%s' available.", root_frame, measured_frame)
 
         if test.robot_config != None and 'wait_for_diagnostics' in test.robot_config and test.robot_config["wait_for_diagnostics"]:
             rospy.loginfo("Waiting for diagnostics to become OK ...")
             r = rospy.Rate(100)
             while not rospy.is_shutdown() and self.diagnostics != None and self.diagnostics.level != 0:
-                rospy.logdebug("... waiting for diagnostics to become OK ...")
+                self.check_for_timeout(start_time, wait_timeout)
```

Comment:
```suggestion
                self.check_for_timeout(start_time, wait_timeout, "latest diagnostics: {}".format(self.diagnostics))
```

Reply:
done in https://github.com/floweisshardt/atf/pull/186

Reply:
cherry-picked relevant changes from #186 and only print latest diagnostic failures to not blow up the log output